### PR TITLE
fix: properly handle shebangs by writing files

### PR DESCRIPTION
**Pull request by kylecarbs**
See https://github.com/coder/coder/pull/10134#discussion_r1351841004

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -1082,6 +1083,7 @@ func (a *agent) handleReconnectingPTY(ctx context.Context, logger slog.Logger, m
 
 		if err = a.trackConnGoroutine(func() {
 			rpty.Wait()
+			cleanup()
 			a.reconnectingPTYs.Delete(msg.ID)
 		}); err != nil {
 			rpty.Close(err)
```
Comment (mafredri) [inline agent/agent.go:1089]: We should cleanup here too.
```suggestion
			rpty.Close(err)
			cleanup()
```

**Review by mafredri — APPROVED** [review]
Looks good, thanks for fixing this! It's been on my todo(w)ish list for ages 😆

File: agent/agentssh/agentssh.go [hunk 2]
```diff
@@ -531,15 +535,31 @@ func (s *Server) CreateCommand(ctx context.Context, script string, env []string)
...
 		}
 		name = words[0]
 		if len(words) > 1 {
 			args = words[1:]
 		} else {
 			args = []string{}
 		}
-		args = append(args, caller, script)
+		scriptSha := sha256.Sum256([]byte(script))
+		tempFile, err := os.CreateTemp("", fmt.Sprintf("coder-script-%x", scriptSha))
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply agent/agentssh/agentssh.go:547]: Reasonable!

**Comment by mafredri** [comment]
I'm going to reopen this one @kylecarbs so it isn't forgotten!

**Comment by mafredri** [comment]
@kylecarbs what do you want to do with this PR? I still think this would be a valuable addition so re-opening for now.

**Pull request closed without merging**